unit/path_strategies.cpp

```diff
@@ -0,0 +1,374 @@
+        symex_eventt::result(symex_eventt::enumt::SUCCESS),
+
+        // Pop line 7 that we saved from above, and execute the loop a second
+        // time. Successful, since this path is infeasible.
+        symex_eventt::resume(symex_eventt::enumt::NEXT, 7),
+        symex_eventt::result(symex_eventt::enumt::SUCCESS),
+
+        // Pop line 7 that we saved from above and bail out. That corresponds to
+        // executing the loop once, decrementing x to 0; assert(x) should fail.
+        symex_eventt::resume(symex_eventt::enumt::JUMP, 9),
+        symex_eventt::result(symex_eventt::enumt::FAILURE),
+      },
+      unwind_limit);
+  }
+}
+
+// In theory, there should be no need to change the code below when adding new
+// test cases...
+
+void symex_eventt::validate_result(
+  listt &events,
+  const safety_checkert::resultt result)
+{
+  REQUIRE(result != safety_checkert::resultt::ERROR);
+
+  if(result == safety_checkert::resultt::SAFE)
+  {
+    REQUIRE(!events.empty());
+    REQUIRE(events.front().first == symex_eventt::enumt::SUCCESS);
+    events.pop_front();
+  }
+  else if(result == safety_checkert::resultt::UNSAFE)
+  {
+    REQUIRE(!events.empty());
+    REQUIRE(events.front().first == symex_eventt::enumt::FAILURE);
+    events.pop_front();
+  }
+}
+
+void symex_eventt::validate_resume(
+  listt &events,
+  const goto_symex_statet &state)
+{
+  REQUIRE(!events.empty());
+
+  if(state.has_saved_jump_target)
+  {
+    REQUIRE(events.front().first == symex_eventt::enumt::JUMP);
+
+    REQUIRE(
+      events.front().second ==
+      std::stoi(state.saved_target->source_location.get_line().c_str()));
+  }
+  else if(state.has_saved_next_instruction)
+  {
+    REQUIRE(events.front().first == symex_eventt::enumt::NEXT);
+
+    REQUIRE(
+      events.front().second ==
+      std::stoi(state.saved_target->source_location.get_line().c_str()));
+  }
+  else
+  {
+    REQUIRE(false);
+  }
+  events.pop_front();
+}
+
+void check_with_strategy(
+  const std::string &strategy,
+  const std::string &program,
+  symex_eventt::listt events,
+  const unsigned unwind_limit)
+{
+  WHEN("strategy is '" + strategy + "'")
+  _check_with_strategy(strategy, program, unwind_limit, events);
+}
+
+// This is a simplified version of bmct::do_language_agnostic_bmc, without all
+// the edge cases to deal with java programs, bytecode loaded on demand, etc. We
+// need to replicate some of this stuff because the worklist is a local variable
+// of do_language_agnostic_bmc, and we need to check the worklist every time
+// symex returns.
+void _check_with_strategy(
+  const std::string &strategy,
+  const std::string &program,
+  const unsigned unwind_limit,
+  symex_eventt::listt &events)
+{
+  temporary_filet tmp("path-explore_", ".c");
+  std::ofstream of(tmp().c_str());
+  REQUIRE(of.is_open());
+
+  of << program << std::endl;
+  of.close();
+
+  register_language(new_ansi_c_language);
+  cmdlinet cmdline;
+  cmdline.set("function", "main");
```

Comment:
Watch out - here if you use the `cmdlinet` the `set` call is ignored as it doesn't think `function` etc is a real flag. I wrote `free_form_cmdlinet` that automatically adds the flags as you set them.

Reply:
@thk123 I don't understand this comment, if I remove that line the test no longer works (it doesn't automatically set the entry point to `main`), so this call must be having some effect?

Reply:
That's surprising, not least because `main` is the default entry point! If easy for you to do, I suggest stepping through `cmdlinet::set` cos I found that it doesn't normally do anything.

Reply:
From reading the code in `cmdline.cpp` I would agree with @thk123, so something weird seems to be happening here. I note that the next line, `cmdline.set("paths", strategy);` is likely redundant as that's all taken care of via `opts.set_option` calls later on. Furthermore it seems that `unit/pointer-analysis/custom_value_set_analysis.cpp` also (successfully?) uses `cmdlinet` in the way it is done here. Insights would be really appreciated!

Reply:
If I remove either `cmdline.set("function", "main");` or `config.main = "main";`, I get
```
Type-checking path-explore_18497.A3WUtT
1: file /tmp/path-explore_18497.A3WUtT.c line 9 function main: function `assert' is not declared
1: main symbol `E.f' not found
```
And everything fails. Do you have any idea why this happens, or should I step through or print out more debug info?

Reply:
`config.set(cmdline);`  should take care of setting `config.main` if the "function" option was set in `cmdline`. So `config.main = "main";` being required is consistent with `cmdline.set` not having any effect. But `cmdline.set("function", "main");` being vital is still a mystery.

Reply:
:sob: of course, it's because I'm slightly distracted at a conference, and so I'm running tests locally on my laptop but modifying the test suite through SSH to my workstation.

I've removed the `set` calls (including the `"paths"` option), thanks for the review. This PR needs rebasing since #2049 was merged, but I'll do that when I'm less distracted...